frontend/src/metabase/lib/expressions/tokenizer.js

```diff
@@ -0,0 +1,328 @@
+              break;
+            case "t":
+              value += "\t";
+              break;
+            case "v":
+              value += "\x0b";
+              break;
+            default:
+              value += seq;
+              break;
+          }
+        }
+      } else {
+        value += ch;
+      }
+    }
+    const type = TOKEN.String;
+    const end = index;
+    const terminated = quote === source[end - 1];
+    const error = terminated ? null : t`Unterminated quoted string`;
+    return { type, value, start, end, error };
+  };
+
+  const scanBracketIdentifier = () => {
+    const start = index;
+    const bracket = source[start];
+    if (bracket !== "[") {
+      return null;
+    }
+    ++index;
+    while (index < length) {
+      const ch = source[index++];
+      if (ch === "]") {
+        break;
+      } else if (ch === "[") {
+        const type = TOKEN.Identifier;
+        const end = index;
+        const error = t`Bracket identifier in another bracket identifier`;
+        return { type, start, end, error };
+      } else if (ch === "\\") {
+        // ignore the next char, even if it's [ or ]
+        index++;
+      }
+    }
+    const type = TOKEN.Identifier;
+    const end = index;
+    const terminated = source[end - 1] === "]";
+    const error = terminated ? null : t`Unterminated bracket identifier`;
+    return { type, start, end, error };
+  };
+
+  const isIdentifierStart = cp => isAlpha(cp) || cp === 0x5f; // underscore;
+
+  const isIdentifierChar = cp =>
+    isAlpha(cp) ||
+    isDigit(cp) ||
+    cp === 0x2e || // dot
+    cp === 0x5f; // underscore
+
+  const scanIdentifier = () => {
+    const start = index;
+    const initial = source.charCodeAt(start);
+    if (!isIdentifierStart(initial)) {
+      return null;
+    }
+
+    while (index < length) {
+      const cp = source.charCodeAt(index);
+      if (!isIdentifierChar(cp)) {
+        break;
+      }
+      ++index;
+    }
+    const end = index;
+    if (index <= start) {
+      return null;
+    }
+    const id = source.slice(start, end).toLowerCase();
+    if (id === OPERATOR.Not || id === OPERATOR.And || id === OPERATOR.Or) {
+      const type = TOKEN.Operator;
+      const op = id;
+      const error = null;
+      return { type, op, start, end, error };
+    }
+    const type = TOKEN.Identifier;
+    const error = null;
+    return { type, start, end, error };
+  };
+
+  const main = () => {
+    const tokens = [],
+      errors = [];
+    while (index < length) {
+      skipWhitespaces();
+      let token = scanOperator();
+      if (!token) {
+        token = scanNumericLiteral();
+      }
+      if (!token) {
+        token = scanStringLiteral();
+      }
+      if (!token) {
+        token = scanIdentifier();
+      }
+      if (!token) {
+        token = scanBracketIdentifier();
+      }
+      if (token) {
+        const { error, ...t } = token;
+        tokens.push(t);
+        if (error && error !== null) {
```

Comment:
if `error` is truthy then it'll never be `null` so I think this can just be `if (error)`